LayoutTests/http/tests/notifications/notification-permission-query.html

```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="/resources/js-test-pre.js"></script>
+<script src="/resources/notifications-test-pre.js"></script>
+</head>
+<body>
+<script>
+
+/*
+ * This test tests that Permissions::query() retrieves the true permission state of the Notifications API. 
+ * Since Notification.permission is implemented correctly, the result of the Permission::query() is compared
+ * against the result of Notification.permission
+ */
+
+jsTestIsAsync = true;
+var Notification_permission, permissions_query;
+
+// Ensure Notification.permission and permission.query return default/prompt by default
+function defaultTest()
+{
+    debug("  Default Test: Test that permissions.query returns prompt by default");
+    debug("  Running test...");
+    navigator.permissions.query({ name: "notifications" }).then((status)=>{
+        Notification_permission = window.Notification.permission;
+        permissions_query = status.state;
+        if (Notification_permission == "default")
+            shouldBeEqualToString("permissions_query", "prompt");
+        else
+            shouldBeEqualToString("Notification_permission", "default");
+        
+        grantTest();
+    });
+}
+
+// Grant the notification permission and ensure Notification.permission 
+// and permission.query return granted
+function grantTest() 
+{    
+    debug("  Grant Test: Test that permissions.query returns granted if permission is granted");
+    internals.withUserGesture(() => {
+        debug("  Granting Permission...");
+        testRunner.grantWebNotificationPermission(testURL);
+
+        debug("  Running Test...");
+        navigator.permissions.query({ name: "notifications" }).then((status)=>{
+            Notification_permission = window.Notification.permission;
+            permissions_query = status.state;
+            shouldBeEqualToString("permissions_query", Notification_permission);
+
+            denyTest();
+        });
+        
+    });
+}
+
+// Deny the notification permission and ensure Notification.permission 
+// and permission.query return denied
+function denyTest() 
+{
+    debug("  Deny Test: Test that permissions.query returns denied if permission is denied");
+    internals.withUserGesture(() => {
+        debug("  Denying Permission...");
+        testRunner.denyWebNotificationPermission(testURL);
+
+        debug("  Running Test...");
+        navigator.permissions.query({ name: "notifications" }).then((status)=>{
+            Notification_permission = window.Notification.permission;
+            permissions_query = status.state;
+            shouldBeEqualToString("permissions_query", Notification_permission);
```

Comment:
For other permissions, we prefer to not provide 'denied' state right away.
Instead, we prefer having prompt and when the page tries to request permission, we return deny for the request and we then expose denied as the permission.
This is a fingerprinting countermeasure that is done for geolocation and microphone/camera IIRC.
Can we apply it there as well?

Reply:
I am not sure I understand the change needs to be made. Are you suggesting the first query would return "prompt" by default until user/client explicitly denies the permission?

query() returns "prompt" by default and `testRunner.denyWebNotificationPermission(testURL);` does the denial process so it should return "denied" here.

Reply:
I see, WebNotificationManager::initialize actually initializes the permission states with WebProcessCreationParameters.
I think this is an existing issue. To fix this we just need to make the initialize() to ignore the denied permissions.

Reply:
Yes, yt does make sense to implement fingerprinting countermeasures in a similar way to camera and microphone. For this pull request, the implementation has been changed to simply enable Permissions::query() to get the permission state of the Notifications API without receiving an error. I believe we've decided to deal with the fingerprinting issue in a later patch.